docs/02-app/01-building-your-application/09-authentication/index.mdx

```diff
@@ -0,0 +1,345 @@
+
+## Authorization
+
+Once a user is authenticated, you’ll need to ensure the user is allowed to visit certain routes and perform operations, such as mutating data with Server Actions, or calling Route Handlers.
+
+### Protecting Routes with Middleware
+
+Middleware in Next.js helps you control who can access different parts of your website. This is important for keeping areas like the user dashboard safe, while letting everyone see public pages like those for marketing. By default, middleware will be applied to all pages. If you anticipate that only certain routes will require authentication or specific processing, you can use the `matcher` property to specify exactly which pages the middleware should run on.
+
+Here's how to implement middleware for authentication in Next.js:
+
+1. **Setting Up Middleware:**
+   - Create a `middleware.ts` or `.js` file in your project's root directory.
+   - Include logic to authorize user access, such as checking for authentication tokens.
+2. **Defining Protected Routes:**
+   - Use the `matcher` option in your middleware to identify routes that require authorization checks.
+3. **Middleware Logic:**
+   - Write logic to verify if a user is authenticated. Check user roles or permissions for route authorization.
+4. **Handling Unauthorized Access:**
+   - Redirect unauthorized users to a login or error page as appropriate.
+
+Example middleware file:
+
+```ts filename="middleware.ts" switcher
+import type { NextRequest } from 'next/server'
+
+export function middleware(request: NextRequest) {
+  const currentUser = request.cookies.get('currentUser')?.value
+
+  if (currentUser) {
+    return Response.redirect(new URL('/dashboard', request.url))
+  }
+  return Response.redirect(new URL('/login', request.url))
+}
+
+export const config = {
+  matcher: '/dashboard',
+}
+```
+
+This example uses [`Response.redirect`](https://developer.mozilla.org/en-US/docs/Web/API/Response/redirect_static) for efficient, centralized access control. It handles redirects early in the request pipeline, making it efficient and ideal for broad strategies.
+
+For specific redirection needs, the `redirect` function can be used in Server Components, Route Handlers, and Server Actions to provide more control. This is useful for role-based navigation or context-sensitive scenarios.
+
+```js
+import { redirect } from 'next/navigation'
+
+export default function Page() {
+  // Logic to determine if a redirect is needed
+  const accessDenied = true
+  if (accessDenied) {
+    redirect('/login')
+  }
+
+  // Define other routes and logic
+}
+```
+
+This flexibility is key when handling redirections based on detailed authentication outcomes or other contextual factors. For example, `redirect('/login')` in a Server Action allows for precise redirection based on the user's role or specific conditions during the authentication process.
```

Comment:
Sorry, I'm not sure I understand this part. What are detailed authentication outcomes and contextual factors? Could we possibly provide use cases?